ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+    doSubscribeTo conThreads targets = do
+        target_m <- getSubscriptionTarget targets
+        case target_m of
+            Just (target, nextTargets) -> do
+                ref <- refConnection tbl target valencyVar
+                case ref of
+                     ConnectionTableCreate -> do
+                        localAddr_m <- getLocalAddress target
+                        case localAddr_m of
+                             Nothing -> -- No usable local address continue with next
+                                        doSubscribeTo conThreads nextTargets
+                             Just localAddr -> do
+                                 -- TODO use asyncWith instead?
+                                 _ <- async (doConnect conThreads localAddr target)
+
+                                 {- The time to wait depends on available deltaQ information for the
+                                 -- destination address and max/min values from RFC8305.
+                                 -}
+                                 let delay = case connectionAttemptDelay target of
+                                                  Just d  -> min maxConnectionAttemptDelay $
+                                                                 max minConnectionAttemptDelay d
+                                                  Nothing -> defaultConnectionAttemptDelay
+                                 traceWith tracer $
+                                     SubscriptionTraceSubscriptionWaitingNewConnection delay
+                                 threadDelay delay
+                                 doSubscribeTo conThreads nextTargets
+                     ConnectionTableExist -> do
+                         {- We already have a connection to this address. -}
+                         traceWith tracer $ SubscriptionTraceConnectionExist target
+                         doSubscribeTo conThreads nextTargets
+                     ConnectionTableDone -> do
+                        traceWith tracer SubscriptionTraceSubscriptionRunning
+                        return ()
+            Nothing -> do
+                len <- fmap length $ atomically $ readTVar conThreads
+                when (len > 0) $
+                    traceWith tracer $ SubscriptionTraceSubscriptionWaiting len
+
+                -- We wait on the list of active connection threads instead of using an async wait
+                -- function since some of the connections may succed and then should be left
+                -- running.
+                atomically $ do
+                    activeCons  <- readTVar conThreads
+                    unless (null activeCons) retry
+
+                valencyLeft <- atomically $ readTVar valencyVar
+                if valencyLeft == 0
+                   then traceWith tracer SubscriptionTraceSubscriptionRunning
+                   else traceWith tracer SubscriptionTraceSubscriptionFailed
+
+
+    doConnect conThreads localAddr remoteAddr =
+        bracket
```

Comment:
Each of the parts of the backet here is quite large. I'd give them names and move them into a `where` clause.